### PR TITLE
Radio buttons partly done

### PR DESCRIPTION
**Pull request by Kasmadei**
Fixes partially: https://github.com/kbss-cvut/fta-fmea-ui/issues/281

Points that needs to be added/checked.

Add: 
- Operational failure rate

Check:
- How works tree update 

@blcham I would like to discuss it tomorrow. For some reason now update endpoint was not even called :( Need some time to find out why. 
Screenshot 2024-06-05 at 8 54 31 PM

### CODE REVIEW

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 1]
```diff
@@ -35,6 +41,30 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
   const [quantity, setQuantity] = useState<number | undefined>(undefined);
   const [schematicDesignation, setSchematicDesignation] = useState<string | undefined>(undefined);
 
+  const [snsPredictedFailureRate, setSnsPredictedFailureRate] = useState<number | undefined>(undefined);
+  const [snsManuallyDefinedFailureRate, setSnsManuallyDefinedFailureRate] = useState<number | undefined>(undefined);
+  const [selectedRadioButton, setSelectedRadioButton] = useState<string>(RadioButtonType.Predicted);
+
+  const handleManuallyDefinedFailureRateChange = (event) => {
+    const inputValue = event.target.value;
+    const regex = /^[0-9]*\.?[0-9]*$/;
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:50]: For manual value property we need to support not only decimal number in this format but also in a scientific format that includes exponent (e.g. as 1.23E3)

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 2]
```diff
@@ -35,6 +41,30 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
...
+    if (regex.test(inputValue)) {
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:51]: I'm not sure what will happen if the regex fails. Do we just silently ignore it?  If so, create a separate issue to resolve this. 

BTW, will we have manual failure rate probability selected while not having a value ? 
Maybe a workaround is to make probability equal to 1 and mark the selected value as red to indicate it.

**Review by blcham — CHANGES_REQUESTED** [review]
@Kasmadei I would like to have a partial solution ready by Friday lunch; please make the pull request meaningful/merge-able while delegating the problems to new issues.

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 1 above)
Comment (blcham) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:50]: https://github.com/kbss-cvut/fta-fmea-ui/issues/368

**Comment by blcham** [comment]
# Additional info

How it should work for the selection of predicted failure rate:
- modify event1.has-selected-failure-rate = {}
- modify event1.probability = event1.has-selected-failure-rate.value
- persist event1 (PUT /fault-event/event1)

`components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx`

**Comment by blcham** [comment]
@Kasmadei check the title of this issue

**Comment by blcham** [comment]
Relates to https://github.com/kbss-cvut/fta-fmea-ui/issues/377

**Comment by Kasmadei** [comment]
@blcham 

When attempting to update a FaultEvent using the update function, a RangeError: Maximum call stack size exceeded is encountered during the axiosClient.put call. This issue appears to be related to circular references within the faultEvent object.
The application throws a RangeError: Maximum call stack size exceeded error during the axiosClient.put request:

```
RangeError: Maximum call stack size exceeded
    at merge (axios.js?v=802ac8ba:105:15)
    at assignValue (axios.js?v=802ac8ba:113:27)
    at forEach (axios.js?v=802ac8ba:82:10)
    at merge (axios.js?v=802ac8ba:121:21)
    at assignValue (axios.js?v=802ac8ba:113:27)
    at forEach (axios.js?v=802ac8ba:82:10)
    at merge (axios.js?v=802ac8ba:121:21)
    at assignValue (axios.js?v=802ac8ba:113:27)
    at forEach (axios.js?v=802ac8ba:82:10)
    at merge (axios.js?v=802ac8ba:121:21)
```
I have never see something like this. I'll try to resolve it

**Comment by blcham** [comment]
Current state:
- operational failure rate is not yet visualized correctly in FTA tree (missing "O")
- server update is not working due to circular dependencies
  - need to check that FTA tree is refreshed after the update of fault event is done

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 3]
```diff
@@ -38,12 +51,69 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
...
+  };
+
   const handleFailureModeClicked = (failureMode: FailureMode) => {
     setFailureModeOverview(failureMode);
     setFailureModeOverviewDialogOpen(true);
   };
 
   useEffect(() => {
+    // Clear values, after node was changed
+    setSnsPredictedFailureRate(undefined);
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:110]: Do we really need to do this?

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 4]
```diff
@@ -21,7 +22,19 @@ interface Props {
   rootIri?: string;
 }
 
+enum RadioButtonType {
+  Predicted = "Predicted",
+  Manual = "Manual",
+  Operational = "Operational",
+}
+
+enum ManualFailureRateType {
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:31]: ? this is  strange name of the enum

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 5]
```diff
@@ -21,7 +22,19 @@ interface Props {
...
+  Sns = "Sns",
+  External = "External",
+}
+
 const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }: Props) => {
+  if (shapeToolData) console.log("shapeToolData", shapeToolData);
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:37]: this console.log does not belong to the PR

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 6]
```diff
@@ -99,11 +169,66 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
...
+      } else if (iriOfSelectedValue === operationalIri) {
+        setSelectedRadioButton(RadioButtonType.Operational);
+      }
+      setSnsManuallyDefinedFailureRate(shapeToolData?.probability);
+    } else {
+      // NO SELECTED ESTIMATE => MANUAL IS SELECTED
+      setSelectedRadioButton(RadioButtonType.Manual);
+      if (shapeToolData?.probability) {
+        setSnsManuallyDefinedFailureRate(shapeToolData.probability);
+        setExternalManuallyDefinedFailureRate(shapeToolData.probability);
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:198]: I do not understand this ... why we have `setExternalManuallyDefinedFailureRate(shapeToolData.probability); ` in the `else` block if for external event predicted and operational failure rates do not exist.

Moreover, we do not have any basicManuallyDefinedFailureRate

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 7]
```diff
@@ -143,35 +170,65 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
...
   const basedFailureRate = shapeToolData?.supertypes?.supertypes?.hasFailureRate?.estimate?.value;
   const requiredFailureRate = shapeToolData?.supertypes?.hasFailureRate?.requirement?.upperBound;
 
+  const FailureRateBox = ({ value, label, rate, selected, classes }) => (
+    <Box display="flex" flexDirection="row" alignItems="center">
+      <FormControlLabel
+        value={value}
+        control={<Radio className={classes.black} />}
+        label={`${label}:`}
+        className={selected ? classes.black : classes.grey}
```
Comment (blcham) [inline src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:226]: We need to have semantic classes according to https://github.com/kbss-cvut/fta-fmea-ui/pull/367 and not colors directly injected here

File: src/components/editor/faultTree/shapes/RenderTree.tsx [hunk 8]
```diff
@@ -58,11 +58,14 @@ const renderTree = async (container, node, parentShape = null, pathsToHighlight)
   // For now it seems impossible to detect when operational failure rate was selected.
   // "has" function from lodash can only check property existence with direct path. So we can't show "(o)" in front of "probability"
 
+  // if (has(node, "probability")) {
+  //   nodeShape.attr(
+  //     ["probabilityLabel", "text"],
+  //     `${has(node, "selectedEstimate") ? "(p)" : "(m)"}${node.probability.toExponential(2)}`,
+  //   );
+  // }
```
Comment (blcham) [inline src/components/editor/faultTree/shapes/RenderTree.tsx:66]: We need to create a separate issue for that.

**Comment by blcham** [comment]
@LaChope please have a look on PR

File: public/locales/cs/translation.json [hunk 9]
```diff
@@ -94,7 +94,11 @@
     "partNumber": "Číslo součásti",
     "stock": "Zásoba",
     "quantity": "Množství",
-    "schematicDesignation": "Schématické označení"
+    "schematicDesignation": "Schématické označení",
```
Comment (grotskat) [inline public/locales/cs/translation.json:97]:
```suggestion
    "schematicDesignation": "Schematické označení",
```

File: public/locales/cs/translation.json [hunk 10]
```diff
@@ -94,7 +94,11 @@
     "partNumber": "Číslo součásti",
```
Comment (grotskat) [inline public/locales/cs/translation.json:94]:
```suggestion
    "partNumber": "Číslo dílu",
```

File: public/locales/cs/translation.json [hunk 11]
```diff
@@ -94,7 +94,11 @@
...
     "stock": "Zásoba",
```
Comment (grotskat) [inline public/locales/cs/translation.json:95]:
```suggestion
    "stock": "Skladem",
```

File: public/locales/cs/translation.json [hunk 12]
```diff
@@ -94,7 +94,11 @@
...
     "quantity": "Množství",
```
Comment (grotskat) [inline public/locales/cs/translation.json:96]:
```suggestion
    "quantity": "Počet",
```

**Review by grotskat — CHANGES_REQUESTED** [review]

File: public/locales/cs/translation.json [hunk 13]
```diff
@@ -94,7 +94,11 @@
...
-    "schematicDesignation": "Schématické označení"
+    "schematicDesignation": "Schématické označení",
+    "requiredFailureRate": "Požadovaná intenzita poruch",
```
[the remark this answers is not in the document]
Comment (grotskat) [reply public/locales/cs/translation.json:98]: @blcham checked

**Comment by blcham** [comment]
@Kasmadei translation are fixed, commited and pushed

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 2 above)
Comment (blcham) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:68]: Will be resolved by https://github.com/kbss-cvut/fta-fmea-ui/issues/407

**Comment by blcham** [comment]
@Kasmadei please process the comments

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx [hunk 14]
```diff
@@ -38,12 +51,69 @@ const FaultEventMenu = ({ shapeToolData, onEventUpdated, refreshTree, rootIri }:
   const [schematicDesignation, setSchematicDesignation] = useState<string | undefined>(undefined);
   const [selectedSystem] = useSelectedSystem();
 
+  const [snsOperationalFailureRate, setSnsOperationalFailureRate] = useState<number | undefined>(undefined);
+  const [snsPredictedFailureRate, setSnsPredictedFailureRate] = useState<number | undefined>(undefined);
+  const [snsManuallyDefinedFailureRate, setSnsManuallyDefinedFailureRate] = useState<number | undefined>(undefined);
+  const [externalManuallyDefinedFailureRate, setExternalManuallyDefinedFailureRate] = useState<number | undefined>(
```
[the remark this answers is not in the document]
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:57]: Yes, there 2 states, "externalManuallyDefinedFailureRate" for external and "snsManuallyDefinedFailureRate" for basic

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 3 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:110]: Yes. We want to be sure, that after node selection, we don't have any saved props from previous node

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 4 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:31]: Agree, looks strange. Renamed to "NodeTypeWithManualFailureRate"

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 7 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:226]: Renamed to selected and notSelected

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 6 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:198]: created a sep. issue for this one: https://github.com/kbss-cvut/fta-fmea-ui/issues/409

File: src/components/editor/faultTree/shapes/RenderTree.tsx (hunk 8 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/shapes/RenderTree.tsx]: New issue: https://github.com/kbss-cvut/fta-fmea-ui/issues/410

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (blcham) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:57]: so why do you call it then snsManuallyDefinedFailureRate ??? I suggest to rename it:
snsManuallyDefinedFailureRate --> basicManuallyDefinedFailureRate

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 7 above)
Comment (blcham) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:226]: where did you rename it? You do not have to necessarily do separate issue here but put here the PR where you solved it

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 5 above)
Comment (blcham) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:37]: @Kasmadei any solution for that?

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 7 above)
Comment (LaChope) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:226]: When possible, I would try to use [MUI default styling variants](https://mui.com/material-ui/api/typography/) using the `variant` prop.

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:57]: Will be renamed in next PR

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 5 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:37]: Was added by mistake (for debug). Removed.

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 7 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:226]: Was replaced with classes: "selected", "notSelected", "editableValue", "notEditableValue"

File: src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx (hunk 4 above)
Comment (Kasmadei) [reply src/components/editor/faultTree/menu/faultEvent/FaultEventMenu.tsx:31]: @blcham For better understanding i'll rename it to "Basic"
